File: services/client/src/protocol/protocol.go
package protocol

import (
	"errors"
	"net"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/protocol/packet"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

type Protocol struct {
	conn net.Conn
}

func NewProtocol(conn net.Conn) Protocol {
	return Protocol{conn: conn}
}

func (p *Protocol) Handshake(agencyId uint8) error {
	if err := p.sendHello(agencyId); err != nil {
		return err
	}

	if err := p.recvAck(); err != nil {
		return err
	}
	return nil
}

func (p *Protocol) sendHello(agencyId uint8) error {
	return p.sendPkt(packet.NewHelloPacket(agencyId))
}

func (p *Protocol) SendBets(bets []lottery.Bet) error {
	if err := p.sendPkt(packet.NewBetPacket(bets)); err != nil {
		return err
	}
	return p.recvAck()
}

func (p *Protocol) recvAck() error {
	pkt, err := p.recvPkt()
	if err != nil {
		return err
	}

	if _, ok := pkt.(*packet.AckPacket); !ok {
		return errors.New("expected ack packet")
	}

	return nil
}

func (p *Protocol) RecvWinners() ([]lottery.Bet, error) {
	winners := make([]lottery.Bet, 0)

	for {
		pkt, err := p.recvPkt()
		if err != nil {
			return nil, err
		}

		switch v := pkt.(type) {
		case *packet.BetPacket:
			winners = append(winners, v.Bets...)
		case *packet.FinPacket:
			return winners, nil
		default:
			return nil, errors.New("unkwnown packet type")
		}
	}
}

func (p *Protocol) SendFin() error {
	return p.sendPkt(packet.NewFinPacket())
}

func (p *Protocol) sendPkt(pkt packet.Packet) error {
	frame, err := pkt.ToBytes()
	if err != nil {
		return err
	}
	return safe_socket.SendAll(p.conn, frame)
}

func (p *Protocol) recvPkt() (packet.Packet, error) {
	headerBytes, err := safe_socket.RecvAll(p.conn, packet.HEADER_LENGTH)
	if err != nil {
		return nil, err
	}

	header, err := packet.ParseHeader(headerBytes)
	if err != nil {
		return nil, err
	}

	payload, err := safe_socket.RecvAll(p.conn, int(header.PayloadLen))
	if err != nil {
		return nil, err
	}

	return packet.FromBytes(header.OpCode, payload)
}

func (p *Protocol) Close() {
	p.conn.Close()
}

File: services/server/src/protocol/protocol.py
import socket

from .packet import AckPacket, FinPacket, BetPacket, HelloPacket, Packet, HEADER_LENGTH, parse_header, from_bytes
from lottery import Bet
from safe_socket import safe_socket

class EndOfBets(Exception):
    """all bets have been received"""

class ConnectionClosed(Exception):
    """the peer closed the connection without sending a Fin packet"""

class Protocol:
    def __init__(self, client_socket: socket.socket):
        self.socket = client_socket
        self.agency_id: int | None = None

    def recv_hello(self) -> int:
        packet = self.__recv_pkt__()
        if not isinstance(packet, HelloPacket):
            raise Exception("expected hello packet")
        self.agency_id = packet.agency_id
        return packet.agency_id

    def send_fin(self) -> None:
        self.__send_pkt__(FinPacket())

    def send_ack(self) -> None:
        self.__send_pkt__(AckPacket())

    def send_winners(self, bets: list[Bet]) -> None:
        if not bets:
            return
        self.__send_pkt__(BetPacket(bets))

    def recv_batch(self) -> list[Bet]:
        packet = self.__recv_pkt__()

        if isinstance(packet, BetPacket):
            return packet.bets
        elif isinstance(packet, FinPacket):
            raise EndOfBets
        else:
            raise Exception("unknown packet type")

    def close(self) -> None:
        self.socket.close()


    def __send_pkt__(self, pkt: Packet) -> None:
        frame = pkt.to_bytes()
        return safe_socket.send_all(self.socket, frame)

    def __recv_pkt__(self) -> Packet:
        buff = safe_socket.recv_all(self.socket, HEADER_LENGTH)
        if not buff:
            raise ConnectionClosed

        header = parse_header(buff)
        payload = safe_socket.recv_all(self.socket, header.length)
        if not payload and header.length > 0:
            raise ConnectionClosed

        return from_bytes(header.opcode, payload, self.agency_id)
